### PR TITLE
refactor: improve the partition compute

### PR DESCRIPTION
**Pull request by ShiKaiWi**
## Rationale
The previous implementation of computing partition key is coarse, leading to poor performance.

## Detailed Changes
Refactor the implementation of computing partition key, and the most of memory allocation is avoided.

## Test Plan
Existing unit tests.

### CODE REVIEW

File: common_types/src/datum.rs [hunk 1]
```diff
@@ -1077,6 +1077,30 @@ pub enum DatumView<'a> {
     Time(i64),
 }
 
+impl<'a> From<&'a Datum> for DatumView<'a> {
```
Comment (jiacai2050) [inline common_types/src/datum.rs:1080]: L941 already have this fn.

File: table_engine/src/partition/rule/key.rs [hunk 2]
```diff
@@ -242,33 +231,130 @@ fn expand_partition_keys_group(
...
+}
+
+impl<'a, T> std::io::Read for PartitionKeysReadAdapter<'a, T>
+where
+    T: Iterator<Item = DatumView<'a>>,
+{
+    /// This implementation will fill the whole buf as much as possible, and the
+    /// only scenario where the buf is not fully filled is that the serialized
+    /// bytes from the `self.key_views` are exhausted.
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
```
Comment (jiacai2050) [inline table_engine/src/partition/rule/key.rs:325]: Add some UT to ensure read is expected.

File: components/hash_ext/src/lib.rs [hunk 3]
```diff
@@ -53,13 +53,13 @@ mod test {
...
     fn test_murmur_hash() {
-        assert_eq!(hash64(&[]), 0);
+        assert_eq!(hash64(&(vec![])[..]), 0);
 
         for (key, code) in [
             (b"cse_engine_hash_mod_test_bytes1", 6401327391689448380),
             (b"cse_engine_hash_mod_test_bytes2", 10824100215277000151),
         ] {
-            assert_eq!(code, hash64(key));
+            assert_eq!(code, hash64(&key[..]));
```
Comment (jiacai2050) [inline components/hash_ext/src/lib.rs:62]: This syntax is ugly, maybe `key.as_slice()` is enough?

**Comment by tanruixiang** [comment]
LGTM

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**